src/libraries/UniswapV3Lib.sol

```diff
@@ -0,0 +1,259 @@
+pragma solidity ^0.8.21;
+
+import { IERC20 } from "openzeppelin-contracts/contracts/interfaces/IERC20.sol";
+import { IERC20Metadata } from "openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+import { IALMProxy }   from "../interfaces/IALMProxy.sol";
+import { IRateLimits } from "../interfaces/IRateLimits.sol";
+
+import { RateLimitHelpers } from "../RateLimitHelpers.sol";
+
+import { FullMath } from "lib/dss-allocator/src/funnels/uniV3/FullMath.sol";
+import { UniV3Utils } from "lib/dss-allocator/test/funnels/UniV3Utils.sol";
+
+import { ISwapRouter, IUniswapV3PoolLike, INonfungiblePositionManager } from "../interfaces/UniswapV3Interfaces.sol";
+
+import { TickMath } from "lib/dss-allocator/src/funnels/uniV3/TickMath.sol";
+
+library UniswapV3Lib {
+    uint24 public constant MAX_TICK_DELTA = 887272; // From https://github.com/sky-ecosystem/dss-allocator/blob/dev/src/funnels/uniV3/TickMath.sol#L15
+
+    uint256 internal constant Q192 = 2 ** 192;
+
+    /**********************************************************************************************/
+    /*** Structs                                                                                ***/
+    /**********************************************************************************************/
+    struct UniswapV3PoolParams {
+        uint24 swapMaxTickDelta;
+        uint32 swapTwapSecondsAgo;
+    }
+
+    struct UniV3Context {
+        IALMProxy   proxy;
+        IRateLimits rateLimits;
+        bytes32     rateLimitId;
+        address     pool;
+        uint256     deadline;
+    }
+
+    struct SwapParams {
+        UniswapV3PoolParams poolParams;
+        address             router;
+        address             tokenIn;
+        uint256             amountIn;
+        uint256             minAmountOut;
+        uint24              tickDelta; // The maximum that the tick can move by after completing the swap; type(uint24).max for no limit
+        uint256             maxSlippage;
+    }
+
+    struct SwapCache {
+        IUniswapV3PoolLike pool;
+        address            token0;
+        address            token1;
+        address            tokenOut;
+        uint160            sqrtPriceLimitX96;
+        uint256            twapExpectedOut;
+    }
+
+    /**********************************************************************************************/
+    /*** External functions                                                                     ***/
+    /**********************************************************************************************/
+
+    // Rate limit decreased by value of token1 
+    function swap(UniV3Context calldata context, SwapParams calldata params) external returns (uint256 amountOut) {
+        SwapCache memory cache = _populateSwapCache(context, params);
+
+        require(
+            params.tokenIn == cache.token0 || params.tokenIn == cache.token1,
+            "UniswapV3Lib/invalid-token-pair"
+        );
```

Comment:
I feel like understanding the context of this would be easier in the `populateSwapCache` function right after it assigns `token0` in `token1`.